File: perm.php
<?php

function Fuchs_perm($buf){
  $res = implode("",$buf);
  echo("$res\n");
  $string_length = count($buf);
  for($i=0;$i<$string_length;$i++){
    $p[$i]=0;
  }
  $i=1;
  while($i < $string_length){
   if ($p[$i] < $i) {
      $j = $i % 2 * $p[$i];
      $tmp=$buf[$j];
      $buf[$j]=$buf[$i];
      $buf[$i]=$tmp;
      $res = implode("",$buf);
      echo("$res\n");
      $p[$i]++;
      $i=1;
   } else {
      $p[$i] = 0;
      $i++;
   }
  }
}


function permutations ($alphabet, $output_length=1) {

    $output = array();

    if ($alphabet AND ($output_length > 0)) {

      // Handles both string alphabets and array alphabets
      if (is_string ($alphabet)) {
        $alphabet_length = strlen ($alphabet);
        $symbol = str_split ($alphabet);
      } elseif (is_array ($alphabet)) {
        $alphabet_length = count ($alphabet);
        $symbol = $alphabet;
      } else {
        return $output;
      }

      if ($alphabet_length < 2) return $output;


      // Creates a -1 index in order to avoid the out-of-bounds
      // warning during the last loop of the do-while structure
      $pointer = array_fill (-1, $output_length+1, 0);
      
      // How much iterations to perform
      $iterations = pow ($alphabet_length, $output_length);
  
      // To avoid all the "- 1"...
      $alphabet_length--;
      $output_length--;
  
      // Do the job
      for ($i=0; $i < $iterations; $i++) {
        $permutation = "";
        for ($c = 0; $c <= $output_length; $c++) {
          $permutation .= $symbol[$pointer[$c]];
        }
        $output[] = $permutation;
  
  
        // Updates the pointers
        $c = $output_length;
  
        do {
          $pointer[$c]++;
          if ($pointer[$c] <= $alphabet_length) {
            break;
          } else {
            $pointer[$c] = 0;
            $c--;
          }
        } while (TRUE);
      }
    }
    
    return $output;
  }

//$start = time();
//Fuchs_perm(range('a', 'd'));
//echo "\nFuchs disposizioni senza ripetizione: ".(time() - $start)."\n";

$alpha = array_merge(range('a', 'z'), range('A', 'Z'), range(0, 9), array('!', '?', '.'));
print_r($alpha);
$start = time();
permutantions($alpha, 4);
echo "\npermutations disposizioni con ripetizione: ".(time() - $start)."\n";
?>

File: perm.cpp
#include <iostream>
#include <string>
#include <math.h>
#include <cstdlib>
#include <thread>
#include <future>

using namespace std;

//4.3 => 4 / 3
int permutations(string alphabet, int output_length, int offset) {
	int iterations = pow(alphabet.length(), output_length-1);
	string* output = new string[iterations];
	int oi = 0; 
	if(alphabet.length() && output_length > 0) {
		if (alphabet.length() < 2) 
			return 0;
			
		int pointer[output_length];
		std::fill(pointer, pointer + output_length, 0);
		pointer[0] = offset;
		
		int alphabet_length = alphabet.length();
      	output_length--;
  
      	for (int i = 0; i < iterations; i++) {
        	string permutation = "";
        	int c;
        	for (c = 0; c <= output_length; c++) {
          		string tmp(1, alphabet.at(pointer[c]));
          		permutation += tmp;
        	}
        	
			string cmd = string("unrar t -y -p")+permutation+string(" ./a.rar 1>/dev/null 2>&1");
			if(system(cmd.c_str()) == 0) {
				string echo = string("echo \"")+permutation+string("\" > /tmp/password");
				cout<<"La password è: "<<permutation;
				system(echo.c_str());
				return 0;
			}

	  
		    c = output_length;
	  
		    do {
				pointer[c]++;
				if (pointer[c] < alphabet_length) {
		        	break;
		      	} else {
		        	pointer[c] = 0;
		        	c--;
		      	}
		    } while (true);
		}
	}
	return 1;
}

int main() {

	string a = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
	int t = a.length();
	thread pool[t];
	for(int d = 1; d < 10; d++) {
		for(int i = 0; i < t; i++) {
			pool[i] = thread(permutations, a, 3, i);
		}
		for(int i = 0; i < t; i++) {
			pool[i].join();
		}
	}
	//thread first(permutations, a, 3, 2, 0);
	//thread second(permutations, a, 3, 2, 1);
	//first.join();
	//second.join();
	return 0;
}
